url79d4f7c7ed_schanq_Live_Algorithm_tgz-pJ8-algorithms_and_systems_AI_Evolutionary_GeneticAlgorithmJ8/src/algorithms_and_systems_AI_Evolutionary/GeneticAlgorithm.java

```diff
@@ -0,0 +1,504 @@
+	// input Chromosome is in comparison to the target			=
+	// ==========================================================
+	public void calculateFitness(Chromosome c){
+		
+		// Determines the Chromosomes fitness according to its Key
+		// and scale genes. As the Scale gene can only be determined
+		// as fit if the Key gene is fit The Level of tonality function is 
+		// used in conjunction with the fitness to determine whether
+		// the chromosome's Key/Scale genes are dominant or recessive.
+		Key [] relatedKeys = this.targetTonality.getRelatedKeys();
+		for(Key k : relatedKeys){
+			if(c.getKeyGene() == k){
+				c.setFitness(4);
+				c.setLevelOfTonality(1);
+				for(Scale s :k.getChord().getScales()){
+					if(c.getScaleGene() == s){
+						c.setFitness(8);
+						c.setLevelOfTonality(2);
+					}
+				}
+				break;
+			}
+		}
+		
+		// Determines whether the Dynamic genes are fit in
+		// comparison to the target chromosome
+		Dynamic [] dyns = this.targetLoudness.allDynamics; 
+		int [] fits     = this.targetLoudness.howCloseAreTheDynamics;
+		DynamicShape ds = this.targetLoudness.getTheShape();
+		
+		for(int i = 0; i < dyns.length; i++){
+			if(dyns[i] ==  c.getDynamicGene()){
+				c.setFitness(c.getFitness()+fits[i]);
+				break;
+			}
+		}
+		if(c.getDshapeGene() == ds){
+			c.setFitness(c.getFitness()+8);
+		}
+		
+		// Determines the Fitness of the Chromosome's rhythmGenes by getting the
+		// beat value of each rhythmic block (considered fit by the target feel) 
+		// and assigning a fitness level for each based on its length so if length 1 then
+		// fitness 2 if beatlength 2 fitness 4 (The purpose is so that the ideal rhythmic 
+		// gene will equate to a fitness of 8). It then compares the chromosome's 
+		// rhythmic_blocks to the ones considered fit, if one matches then add the 
+		// appropriate fitness value to the chromosome's fitness.
+		ArrayList <Rhythmic_Block> fitRhythms = this.targetFeel.getFitRhythms();
+		int [] levelsOfFitness = new int [fitRhythms.toArray().length];
+		for(int i = 0; i < levelsOfFitness.length; i++){
+			if(fitRhythms.get(i).getBeatValue() == 1){
+				levelsOfFitness[i] = 2;
+			}
+			else if(fitRhythms.get(i).getBeatValue() == 2){
+				levelsOfFitness[i] = 4;
+			}
+		}
+		int i = 0;
+		for(Rhythmic_Block rb : fitRhythms){
+			for(Rhythmic_Block r : c.rhythmGenes){
+				if(r.equals(rb)){
+					c.setFitness(c.getFitness()+levelsOfFitness[i]);
+				}
+			}
+			i++;
+		}
+	}
+	// ==========================================================
+	// Genetic crossover - crosses over the genes, however the  =	
+	// specific difference between this and other crossover		=
+	// functions is that it is slightly hard coded so that the 	=
+	// tonal genes (Key and Scale) are crossed over together if =
+	// they are fit.											=
+	// ==========================================================
+	public Chromosome crossover(Chromosome mother, Chromosome father){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 39. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 255-264, lines 274-285, lines 287-338 into separate methods.